encoding/masking-limiting-etc.md

```diff
@@ -83,22 +126,107 @@ Internally it uses a mask called a Camembert to generate a larger mask
+[DeHalo_alpha][dehalo_alpha].
+
+However, we must be cautious in applying that filter,
+since, while removing halos reliably,
+it’s extremly destructive to the lineart as well.
+Therefore we must use a **dehalo mask**
+to protect the lineart and limit the filtering to halos.
+
+
+A dehalo mask aims to cover the halos
+but exclude the lines themselves,
+so that the lineart won’t be blurred or dimmed.
+In order to do that,
+we first need to generate an edgemask.
+In this example,
+we’ll use the built-in Sobel function.
+After generating the edge mask, we extract the luma plane:
+
+```py
+mask = core.std.Sobel(src, 0)
+luma = core.std.ShufflePlanes(mask, 0, colorfamily=vs.GRAY)
+
+```
+
+![mask luma](images/luma0.png)
+
+Next, we expand the mask twice, so that it covers the halos. 
+``vsutil.iterate`` is a function in [vsutil][vsutil iterate]
+which applies the specified filter a specified number of times to a clip – 
+in this case it runs ``std.Maximum`` 2 times.
 
 ```py
-#black line mask
-mask = zz.dhhmask(src)
+mask_outer = vsutil.iterate(luma, core.std.Maximum, 2)
+```
 
-#grow mask to cover halos
-mask_outer = kgf.iterate(mask, core.std.Maximum, 3)
+![mask_outer](images/mask_outer0.png)
+Now we shrink the expanded clip back
+to cover only the lineart. 
+Applying ``std.Minimum`` twice
+would shrink it back to the edge mask’s original size, 
+but since the edge mask covers part of the halos too, 
+we need to erode it a little further. 
+
+The reason we used ``mask_outer`` as the basis and shrank it thrice,
+instead of using ``mask`` and shrinking it once, 
+which would result in a similiar outline,
+is that this way,
+small adjacent lines with gaps in them (i.e areas of fine texture or details), 
+such as the man’s eyes in this example,
+are covered up completely,
+preventing detail loss. 
 
-#extend a bit and shrink back to exclude close, adjacent lines
-mask_inner = core.std.Maximum(mask_outer)
-mask_inner = kgf.iterate(mask_inner, core.std.Minimum ,4)
+```py
+mask_inner = vsutil.iterate(mask_outer, core.std.Minimum, 3)
+```
 
+![mask_inner](images/mask_inner0.png)
+Now we substract the outer mask covering the halos and the lineart from the inner mask covering only the lineart. 
+This yields a mask covering only the halos,
+which is what we originally wanted:
+
+```py
 halos = core.std.Expr([mask_outer, mask_inner], 'x y -')
 ```
 
+![halos](images/halos0.png)
+Next, we do the actual dehaloing:
+
+```py
+dehalo = hf.DeHalo_alpha(src)
+```
+
+![dehalo](images/dh0.png)
+Lastly, we use MaskedMerge to merge only the filtered halos into the source clip,
+leaving the lineart mostly untouched:
+
+```py
+masked_dehalo = core.std.MaskedMerge(src, dehalo, halos)
+```
+
+![voilá: masked_dehalo](images/dehalod0.png)
+
+[comparison]: https://slowpics.org/comparison/96cbeca4-b4be-4dfc-82b1-631bbc85cdb0
+[dehalo_alpha]: https://github.com/HomeOfVapourSynthEvolution/havsfunc/blob/master/havsfunc.py#L393
```

Comment:
Linking to lines in master branches may lead to the line numbers changing as the code is updated. Although permalinking older revisions might point to outdated versions, it's still better as the reference will continue to work properly when the source is updated.

You can retrieve a permalink with the latest revision of a file by pressing `y` while browsing Github.

Reply:
Uh, you changed this for `vsutil iterate` but not for `dehalo_alpha`.

Reply:
Oops. Fixed.